### PR TITLE
Feature/refactor warmup

### PR DESCRIPTION
**Pull request by Verbalist**

### CODE REVIEW

File: core/repository/block/index.ts [hunk 1]
```diff
@@ -13,112 +10,54 @@ export interface IBlockRepository extends IBlockRepositoryShared {
...
 
     public getLastBlock(): Block {
-        return this.lastBlock;
+        return this.memoryBlocks.length ? this.memoryBlocks[this.memoryBlocks.length - 1] : null;
     }
 
     public deleteLastBlock(): Block {
-        this.memoryBlocks.length = this.memoryBlocks.length - 1;
-        this.lastBlock = this.memoryBlocks[this.memoryBlocks.length - 1];
+        this.memoryBlocks.pop();
```
Comment (axelsheva) [inline core/repository/block/index.ts]: Should we be able to remove a genesis block?

Comment (Verbalist) [reply core/repository/block/index.ts]: Yes, function in repository independent of logic (like genesisBlock)

File: core/service/round.ts [hunk 2]
```diff
@@ -112,51 +117,93 @@ class RoundService implements IRoundService {
...
+            // this.isRoundChainRestored = true;
+            await this.generateRound(block.createdAt);
             return;
         }
+
         const currentRound = RoundRepository.getCurrentRound();
         const firstSlot = RoundRepository.getFirstSlotInRound();
         const lastSlot = RoundRepository.getLastSlotInRound();
-
+        console.log('block', SlotService.getSlotTime(firstSlot), block.createdAt, SlotService.getSlotTime(lastSlot));
```
Comment (axelsheva) [inline core/service/round.ts]: console.log

File: core/service/round.ts [hunk 3]
```diff
@@ -112,51 +117,93 @@ class RoundService implements IRoundService {
...
+            // start forging block at mySlotTime
+            let cellTime = SlotService.getSlotRealTime(mySlot) - new Date().getTime();
+            if (cellTime < 0 && cellTime + MAX_LATENESS_FORGE_TIME >= 0) {
+                cellTime = 0;
+            }
+            if (cellTime >= 0) {
+                logger.info(
+                    `${this.logPrefix}[generateRound] Start forging block to: ${mySlot} after ${cellTime} ms`
+                );
+                console.log('first slot', SlotService.getTime());
```
Comment (axelsheva) [inline core/service/round.ts]: console.log

File: shared/util/config.ts [hunk 4]
```diff
@@ -22,7 +17,13 @@ import testnetGenesisBlock from '../../config/testnet/genesisBlock.json';
 import mainnetGenesisBlock from '../../config/mainnet/genesisBlock.json';
 import { Transaction, IAsset } from 'shared/model/transaction';
 
+const env = process.env;
+
+const validator: Validator = new ZSchema({});
```
Comment (axelsheva) [inline shared/util/config.ts]: Is it used?

Comment (Verbalist) [reply shared/util/config.ts]: +

**Pull request merged**